### PR TITLE
add section on typehints

### LINKED ISSUE #145: [Request] tyephints section
I love this book! This is a fantastic guide for training new research software engineers. I think it would be even better if it had a section in typehints. Would there be any interest in including a section on this?

### PR DESCRIPTION
**Pull request by bendichter**
fix #145

### CODE REVIEW

**Comment by TomasBeuzen** [comment]
Sorry for my radio silence here. Been bogged down with family and work the last few weeks, but I’ll review in the next day or two!

File: py-pkgs/06-documentation.ipynb [hunk 1]
```diff
@@ -458,6 +459,69 @@
...
+    "```\n",
+    "\n",
+    "In this function, `name: str` specifies that `name` should be a string, and `-> str` indicates that the function returns a string.\n",
+    "\n",
+    "For more complex types, the `typing` module provides a wealth of typehints like `List`, `Dict`, `Optional`, and `Union`:\n",
+    "\n",
+    "```python\n",
+    "from typing import List, Optional\n",
+    "\n",
+    "def process_items(items: List[str]) -> Optional[str]:\n",
```
Comment (TomasBeuzen) [inline py-pkgs/06-documentation.ipynb:490]: As `items` is not typed as `Optional`, I suggest we change this example function to the following (I think the intent was that `items` could be an empty list `[]` and hence `if items:` would not pass and return `None`, but to me, the user would have still passed a list so returning a `None` is counter-intuituve):

```
from typing import List, Optional

def process_items(items: List[str]) -> Optional[str]:
    if isinstance(items, list):
        return " ".join(items)
    return None
```

File: py-pkgs/06-documentation.ipynb [hunk 2]
```diff
@@ -458,6 +459,69 @@
...
+   "source": [
+    "(06:Typehints)=\n",
+    "### Typehints"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Typehints in Python are a powerful tool for documentation and readability, enhancing the understanding of code by specifying the expected types of variables, function parameters, and return values. Introduced in Python 3.5 through [PEP 484](https://www.python.org/dev/peps/pep-0484/), typehints offer a way to statically indicate the types within your Python code, enabling tools like type checkers, IDEs, and linters to provide better support for code verification, auto-completion, and refactoring.\n",
```
Comment (TomasBeuzen) [inline py-pkgs/06-documentation.ipynb:474]: Change "enhancing the understanding of code" to ", they help users understand your code by specifying..."

**Review by TomasBeuzen — CHANGES_REQUESTED** [review]
One requested change and one optional one - fantastic writing here!

**Comment by TomasBeuzen** [comment]
@bendichter - when you address PR comments could you please also add your github tag to the **Acknowledgements** in `00-preface.ipynb` (it's in alphabetical order)

File: py-pkgs/06-documentation.ipynb (hunk 1 above)
Comment (bendichter) [reply py-pkgs/06-documentation.ipynb:490]: good points. I'm happy to make this adjustment.

**Comment by bendichter** [comment]
@TomasBeuzen Thanks for the review. I have made all the changes you requested.

**Review by TomasBeuzen — APPROVED** [review]

**Comment by TomasBeuzen** [comment]
Thank you very much for the contribution @bendichter - it is much appreciated! 🙏

**Comment by bendichter** [comment]
Woohoo!

**Pull request merged**